feat: Reconciliation dashboard and Plaid integration

chdecultot opened the pull request:
Dear all,

Please find attached a PR for the redesign of the Bank Reconciliation tool.
This new tool is based on the issue https://github.com/frappe/erpnext/issues/15600
The corresponding doc will be available in the Foundation app.

This feature contains two parts:
1. An integration with Plaid.com to allow the synchronization of bank accounts for users willing to subscribe to Plaid.

  - Addition of a new account
  ![new_account_creation](https://user-images.githubusercontent.com/4903591/49875770-1cc7cf00-fe22-11e8-86d9-5591b1931253.gif)

  - Synchronization of an account
  ![plaid_synchronization](https://user-images.githubusercontent.com/4903591/49875769-1c2f3880-fe22-11e8-9cdc-92154b11df08.gif)


2. An manual upload tool with a reconciliation dashboard

  - Upload of a bank statement
  ![bank_transaction_upload](https://user-images.githubusercontent.com/4903591/49875865-5f89a700-fe22-11e8-8019-178a280aafc9.gif)

  - Semi-auto reconciliation
  ![auto_reconciliation](https://user-images.githubusercontent.com/4903591/49875925-7fb96600-fe22-11e8-93bb-649390bd395d.gif)

  - Manual reconciliation
  ![manual_reconciliation](https://user-images.githubusercontent.com/4903591/49875937-8c3dbe80-fe22-11e8-9e39-4f18dbcf6fe0.gif)


You can also create a new invoice/payment entry or expense claim directly from a bank transaction.

Please let me know if you have any question.

Best regards,

Charles-Henri

chdecultot committed:
Plaid integration

chdecultot committed:
Reconciliation dashboard wip

chdecultot committed:
Bank reconciliation wip

chdecultot committed:
Bank reconciliation wip

chdecultot committed:
Bank reconciliation WIP

chdecultot committed:
WIP

chdecultot committed:
Bank reconciliation dashboard

chdecultot committed:
Cleanup dev

chdecultot committed:
Dev cleanup

chdecultot committed:
Codacy corrections + sql queries

chdecultot commented:
@parsimonyERP

chdecultot committed:
Dev cleanup

chdecultot committed:
Codacy corrections

chdecultot committed:
Codacy corrections

chdecultot committed:
Modify SQL queries and add a test case

chdecultot committed:
Duplicate query to avoid SQL injection

chdecultot committed:
Addition of test cases

chdecultot committed:
Correct test case for Travis

chdecultot committed:
Correct Travis error

rmehta commented:
- [ ] "Link a new bank account" should not be a permanent button
- [ ] Add some empty state. Add a message "Upload a bank statement, or link a bank account"

chdecultot committed:
UX enhancements

chdecultot commented:
Thanks @rmehta for your feedback!

As requested I have changed the following:

- Addition of an empty state:
![image](https://user-images.githubusercontent.com/4903591/49998096-a4811b00-ff93-11e8-8384-d68bcda157f2.png)

- Down caret on every row changed to a more descriptive button:
![image](https://user-images.githubusercontent.com/4903591/49998144-c7133400-ff93-11e8-91b7-9009ad79387e.png)


As for the "Link a new account" button, it already only appears if the plaid integration is activated.
If plaid is not enabled, it doesn't show up (it is similar for the "Synchronize button"):
![image](https://user-images.githubusercontent.com/4903591/49998210-f9249600-ff93-11e8-9bcd-7ed464d71dc9.png)
![image](https://user-images.githubusercontent.com/4903591/49998225-05a8ee80-ff94-11e8-8589-078c9313af16.png)

Let me know if you see  anything else needed additional enhancements!

Thanks a lot!

rmehta commented:
@chdecultot more feedback:

- Too many primary actions! There must be only one primary action in a page
- Give a minimum height and padding around the empty state
- Linking a new account is not an operational action. An even better idea would be to move it to the the "Plaid Settings" page. Also if there are multiple accounts linked, the settings page could be where you could unlink it too.

chdecultot committed:
UX corrections + additional tests

chdecultot committed:
Move actions menu to standard menu

chdecultot committed:
Add button to unlink bank account

chdecultot commented:
Hi @rmehta,

Thanks again for your valuable feedback.
I have modified the UI as requested:

1. All button have been changed to default button (and actions to a standard menu) to leave the primary action color for the "Submit" button (for data upload) or "Reconcile".
![image](https://user-images.githubusercontent.com/4903591/50015101-05294b80-ffc6-11e8-96d0-572839f81c72.png)

![image](https://user-images.githubusercontent.com/4903591/50015131-15d9c180-ffc6-11e8-923c-4c36e706ed5d.png)

![image](https://user-images.githubusercontent.com/4903591/50015225-491c5080-ffc6-11e8-8f46-0bd26479215c.png)


As you can see above, I have added some height and padding around the empty state (similarly to the empty state in query reports)

For Plaid, I have moved the button to add a new bank account to Plaid Settings:
![image](https://user-images.githubusercontent.com/4903591/50015286-7a951c00-ffc6-11e8-9b54-0d2613999f7f.png)

To unlink bank account, I have added a button in the bank account itself with an option to unlink it if it is linked with any external service (Plaid for example).

![image](https://user-images.githubusercontent.com/4903591/50016258-f3957300-ffc8-11e8-9788-73f2f56b660a.png)

chdecultot committed:
Codacy corrections

chdecultot committed:
Codacy corrections

chdecultot committed:
Codacy corrections

chdecultot committed:
Codacy corrections

rmehta commented:
@chdecultot 👍 

@nabinhait please review and merge!

saurabh6790 commented:
This should go in Develop as a Part of next version. Please switch base to develop!

saurabh6790 reviewed (dismissed):
Please change base to develop by rebasing from develop

chdecultot committed:
Codacy correction

chdecultot committed:
Merge conflict resolution

sagarvora commented on erpnext/accounts/doctype/purchase_invoice/purchase_invoice.json:
```diff
@@ -4726,7 +4758,7 @@
  "max_attachments": 0, 
  "menu_index": 0, 
- "modified": "2019-01-07 16:51:59.800081", 
+ "modified": "2018-12-06 09:00:41.508642", 
```
Please use a newer modified timestamp to ensure that the changes you made are updated upon migration.

chdecultot replied:
Sorry I saw 2018 instead of 2019...
Just corrected it.
Thanks for a lot for the feedback!

chdecultot committed:
Purchase invoice modified date

nabinhait committed:
Merge branch 'staging-fixes' into plaid_reconciliation

sagarvora commented:
@chdecultot Pls. resolve merge conflict.
@nabinhait  Pls. review.

chdecultot committed:
Merge branch 'hotfix' into plaid_reconciliation

chdecultot commented:
Hi @sagarvora,

Conflict resolved! :wink:

parsimonyERP commented:
Seems like this PR is close.  Hope it gets some time on the review table soon.

chdecultot committed:
Merge branch 'hotfix' into plaid_reconciliation

chdecultot commented:
Hi,
Please let me know if anything else can be done to help merging this PR.

Thanks!

nabinhait commented:
@chdecultot I have started testing and reviewing the PR and making some changes on the way. It will surely be merged within a week.

chdecultot commented:
Thanks a lot @nabinhait !

Let me know if you need any help (don't hesitate to email me)

parsimonyERP commented:
+1  Thank you Nabin.

chdecultot committed:
Fix: reference doc clearance update correction

rishinikhil commented:
Any timelines on this?

Olawale1 commented:
Hope this gets merged soon... it'll really help a great deal


Thanks

nabinhait commented:
merged manually in hotfix branch, closing this PR

rishinikhil commented:
Hi, when would this be released to SaaS users?

RPPL001 commented:
Manual Upload of bank statement is not working completely. It is uploading only credit transactions but does not upload debit transactions

chdecultot commented:
@RPPL001,

You can configure the upload mapping in the Bank doctype.
Please check the documentation: https://erpnext.com/docs/user/manual/en/accounts/bank-reconciliation

vorasmit commented:
Hi,

We have tested for manual integration and have faced following problems. Could you please look into it.

- Unable to reconcile payments made through Journal Entry. When you select journal entry for reconciliation, you get an error saying entry not found.

- Because we faced this issue, we tried to manually reconcile by directly going into specific bank transaction and linking it. But using this, the "bank transaction date" is not updated in payment entries.

- Check for already imported transactions is not there. If you manually reimport complete bank statement, it should generate new entries only for the new transaction. However, currently, there is duplication of these entries.

- Automatic reconciliation is fantastic. It works well. But it's interface has some scope for improved. Multiple entries should be possibly reconciled in one go, without the requirement to click each transaction, check if automatic reconciliation is available and then click again to reconcile.

Otherwise, It's really great a work.

RPPL001 commented:
Thank you for quick response. Please see my comments in red.


  *   Unable to reconcile payments made through Journal Entry. When you select journal entry for reconciliation, you get an error saying entry not found.
     *   We will request you to identify our mistakes and kindly let us know.
  *   Because we faced this issue, we tried to manually reconcile by directly going into specific bank transaction and linking it. But using this, the "bank transaction date" is not updated in payment entries.
     *   First, we manually uploaded bank statement for month of Jan 2019(which include debit and credit transactions). After upload, system showed only credit transactions. It did not show debit transaction. We matched all the credit transactions. Worked Great.
     *   Second, we uploaded bank statement (only debit Transaction) for Jan 2019. Considering that we might have done some mistakes. System did not show any debit transactions. This is why there may be duplicate entries.
     *   Third, we uploaded bank statements (which includes debit & credit transactions) of Feb 2019. Once again system is showing only credit transactions we are not able to find debit transactions.
  *   Check for already imported transactions is not there. If you manually reimport complete bank statement, it should generate new entries only for the new transaction. However, currently, there is duplication of these entries.
     *   We did it. System tells us 43 transactions uploaded but it does not show credit transactions. As I have mentioned earlier, we have upload bank transaction twice that may created duplicate entries. Once again, please identity and suggest how to see these duplicate entries
  *   Automatic reconciliation is fantastic. It works well. But it's interface has some scope for improved. Multiple entries should be possibly reconciled in one go, without the requirement to click each transaction, check if automatic reconciliation is available and then click again to reconcile.
     *   How we can do it please suggest it. We are will to upgrade.

Thanks,

Loveleet Gera
From: vorasmit <notifications@github.com>
Sent: April 25, 2019 3:40 AM
To: frappe/erpnext <erpnext@noreply.github.com>
Cc: Loveleet Gera <lgera@rpplindustries.com>; Mention <mention@noreply.github.com>
Subject: Re: [frappe/erpnext] feat: Reconciliation dashboard and Plaid integration (#16202)


Hi,

We have tested for manual integration and have faced following problems. Could you please look into it.

  *   Unable to reconcile payments made through Journal Entry. When you select journal entry for reconciliation, you get an error saying entry not found.
  *   Because we faced this issue, we tried to manually reconcile by directly going into specific bank transaction and linking it. But using this, the "bank transaction date" is not updated in payment entries.
  *   Check for already imported transactions is not there. If you manually reimport complete bank statement, it should generate new entries only for the new transaction. However, currently, there is duplication of these entries.
  *   Automatic reconciliation is fantastic. It works well. But it's interface has some scope for improved. Multiple entries should be possibly reconciled in one go, without the requirement to click each transaction, check if automatic reconciliation is available and then click again to reconcile.

Otherwise, It's really great a work.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/frappe/erpnext/pull/16202#issuecomment-486555522>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ALNP34QSYVBZIUA6UK46AT3PSFN6VANCNFSM4GJ6IMGA>.

chdecultot commented:
Hi @vorasmit,

Thanks for your feedback. I just submitted a PR with some corrections.

Regarding your comments:

- Unable to reconcile payments made through Journal Entry. When you select journal entry for reconciliation, you get an error saying resource not available.
**=> That should be fixed with https://github.com/frappe/erpnext/pull/17396**

- Because we faced this issue, we tried to manually reconcile by directly going into specific bank transaction and linking it. But using this, the "bank transaction date" is not updated in payment entries.
**=> That should be fixed too. Now the clearance will be done when saving the bank transaction, not only through the bank reconciliation tool (was a design mistake...)**

- Check for already imported transactions is not there. If you manually reimport complete bank statement, it should generate new entries only for the new transaction. However, currently, there is duplication of these entries.
**=> I do agree that it would be great, but there is a big issue in order to build a tool that imports incrementally: not all banks provide a unique ID each transaction on a csv/xlsx bank statement. Therefore it's complicated to make a uniqueness rule, since it could happen that two transactions have the same characteristics (same date, debit, credit, description...). But it would be a very nice improvement if you could suggest a way to make it work. Maybe for a v2 ?**

- Automatic reconciliation is fantastic. It works well. But it's interface has some scope for improved. Multiple entries should be possibly reconciled in one go, without the requirement to click each transaction, check if automatic reconciliation is available and then click again to reconcile.
**=> I completely agree that it has some scope for improvement, but as discussed with the Core team, the first milestone is to make sure the current tool works as expected. Once all the edge cases have been cleared, the community could pick up the project and work on a v2 with UX and performance improvements.**

Have a nice week-end!

vorasmit commented:
I agree uniqueness cannot be checked by any individual field.

I suggest concatenation of all fields uploaded in a bank statement can be used. It includes the date, description, transaction ref and amount.

And again also true about UX. It's good to break in parts.


Thanks for the fixes and really it's a good job. Keep it up.

albgithub commented:
@chdecultot  and @rmehta Please note the following from Plaid website. Could this be updated? ERPNext is now unable to connect to Plaid with current settings.

A note on Plaid keys

All implementations of Plaid created after July 23, 2020 will use a link_token instead of a public_key. To learn more, view our docs.

If you see a reference to the Plaid public_key outside of the Plaid docs, please let us know by reporting the reference.

parsimonyERP commented:
+1
